Engineering redesign part 4: Bluespace drive part 1

SomeAngryMiner opened the pull request:
What it says on the tin; New BSD machine, with proper ambiance, particles, and the like.

WIP Because there's a single unused var, and the post-explosion effects are not done yet, but should be on a working enough state for the upcoming reason its being PRed earlier.

![image](https://user-images.githubusercontent.com/53237389/168446275-a705c1a7-6a38-42a1-935b-8f400c815307.png)
Animations, sound effects, particles and broken states not included in preview.

:cl: SomeAngryMiner, Beanysprout, Crimsonshrike, Dudezube 
rscadd: Replaces the bluespace drive placeholder with a proper machine, effects and sounds included. Post-explosion effects not included.
/:cl:

SomeAngryMiner committed:
BSD Skeleton

BSD skeleton

Tennessee116 commented on code/game/machinery/bluespace_drive.dm:
```diff
@@ -0,0 +1,149 @@
+	idle_power_usage = 15000
+	construct_state = /decl/machine_construction/default/panel_closed
+
+	machine_name = "bluespace drive"
+	machine_desc = "A bewilderingly complex prototype designed to split the fabric of spacetime and allow access to Bluespace."
```
This is usually different from the regular `name` and `desc`.

SomeAngryMiner replied:
changed.

CrimsonShrike commented on code/game/machinery/bluespace_drive.dm:
```diff
@@ -0,0 +1,149 @@
+	density = TRUE
+	pixel_y = -32
+	pixel_x = -32
+	idle_power_usage = 15000
```
Not really a requirement but you can use the KILOWATT define to make it mroe readable

Spookerton commented on code/game/machinery/bluespace_drive.dm:
```diff
@@ -0,0 +1,149 @@
+	var/drive_sound
+	var/image/overlay
+
+particles/bluespace_torus
```
```suggestion

/obj/item/stock_parts/circuitboard/bluespacedrive
	name = T_BOARD("skrell magic")
	build_path = /obj/machinery/bluespacedrive

/particles/bluespace_torus
```

SomeAngryMiner replied:
put in with the rest of the boards on the circuit folder tree, and added the missing /

MuckerMayhem commented on code/game/machinery/bluespace_drive.dm:
```diff
@@ -0,0 +1,149 @@
+	else
+		icon_state = "bsd_core_broken"
+	set_light(1, 5, 15, 10, COLOR_CYAN)
+	overlay.plane = EFFECTS_ABOVE_LIGHTING_PLANE
```
Causes a roundstart runtime right here:
```
[12:00:00] Runtime in bluespace_drive.dm,55: Cannot modify null.plane.
  proc name: on update icon (/obj/machinery/bluespacedrive/on_update_icon)
  src: the bluespace drive (/obj/machinery/bluespacedrive)
  src.loc: the processing strata (84,98,4) (/turf/simulated/floor/bluegrid)
  call stack:
  the bluespace drive (/obj/machinery/bluespacedrive): on update icon()
  the bluespace drive (/obj/machinery/bluespacedrive): update icon()
  the bluespace drive (/obj/machinery/bluespacedrive): update icon()
  the bluespace drive (/obj/machinery/bluespacedrive): Initialize(1)
  Atoms (/datum/controller/subsystem/atoms): InitAtom(the bluespace drive (/obj/machinery/bluespacedrive), /list (/list))
  Atoms (/datum/controller/subsystem/atoms): InitializeAtoms()
  Atoms (/datum/controller/subsystem/atoms): Initialize(850179)
  Atoms (/datum/controller/subsystem/atoms): DoInitialize(850179)
  Main (/datum/controller/master): Initialize(10, 0)
```

MuckerMayhem reviewed (changes requested):
I think the machine desc and the default desc could be combined so there isn't two separate descriptions:
![image](https://user-images.githubusercontent.com/1161516/168459030-44a7f9d8-4a37-404a-be94-55bf62c9728f.png)

MuckerMayhem commented:
After some more testing, laser shots go underneath the machine itself instead overlaying on top of it. Upon destroying it there's also a runtime in the do_after:
```
[00:35:34] Runtime in mobs.dm,140: undefined variable /obj/machinery/bluespacedrive/var/hand
  proc name: do after detailed (/proc/do_after_detailed)
  src: null
  call stack:
  do after detailed(the bluespace drive (/obj/machinery/bluespacedrive), 40, null, 2378, 61)
  do after(the bluespace drive (/obj/machinery/bluespacedrive), 40, null, 2378, 61)
  the bluespace drive (/obj/machinery/bluespacedrive): take damage(40)
  the bluespace drive (/obj/machinery/bluespacedrive): bullet act(the laser (/obj/item/projectile/beam), "chest")
  the laser (/obj/item/projectile/beam): Bump(the bluespace drive (/obj/machinery/bluespacedrive), 1)
  the processing strata (84,98,4) (/turf/simulated/floor/bluegrid): Enter(the laser (/obj/item/projectile/beam), the deck (84,97,4) (/turf/simulated/floor/tiled/techfloor/grid))
  the laser (/obj/item/projectile/beam): Move(the processing strata (84,98,4) (/turf/simulated/floor/bluegrid), null)
  the laser (/obj/item/projectile/beam): Move(the processing strata (84,98,4) (/turf/simulated/floor/bluegrid), null)
  the laser (/obj/item/projectile/beam): Move(the processing strata (84,98,4) (/turf/simulated/floor/bluegrid))
  the laser (/obj/item/projectile/beam): Process()
```

Lastly, the core effect still persists after destruction, did you mean to have it deleted or is it supposed to stick around?
![image](https://user-images.githubusercontent.com/1161516/168459927-8ff60d8b-c9e6-4dac-9065-d24a8e9cf991.png)

SomeAngryMiner commented:
>Lastly, the core effect still persists after destruction, did you mean to have it deleted or is it supposed to stick around?

Meant to stick around.

>laser shots go underneath the machine itself instead overlaying on top of it.

For this i have an idea how to fix. Will look into it tomorrow.

>I will try and propose some solutions for the runtiming tomorrow.

Thanks lots.

MuckerMayhem commented on code/game/machinery/bluespace_drive.dm:
```diff
@@ -0,0 +1,149 @@
+	if(drive_integrity <= 0)
+		playsound(loc, 'sound/machines/BSD_explosion.ogg', 100)
+		visible_message(SPAN_DANGER(FONT_LARGE("\The [src] beings emitting an ear-splitting, horrible shrill! get back!")))
+		if(do_after(src, 4 SECONDS))
+			visible_message(SPAN_DANGER(FONT_LARGE("\The [src]'s containment field is wracked by a series of horrendous distortions, buckling and twisting like a living thing before bursting in a flash of light!")))
+			explosion(get_turf(src),-1, 5, 10)
+			empulse(get_turf(src),7, 14)
+			drive_broken = TRUE
+			drive_core_unstable = TRUE
+			update_icon()
+		else
+			return
```
```suggestion
		addtimer(CALLBACK(src, .proc/explode), 4 SECONDS)

/obj/machinery/bluespacedrive/proc/explode()
	visible_message(SPAN_DANGER(FONT_LARGE("\The [src]'s containment field is wracked by a series of horrendous distortions, buckling and twisting like a living thing before bursting in a flash of light!")))
	drive_integrity = 0
	explosion(get_turf(src),-1, 5, 10)
	empulse(get_turf(src),7, 14)
	drive_broken = TRUE
	drive_core_unstable = TRUE
	update_icon()
```

do_after isn't really good for this, timer is the best bet.

SomeAngryMiner commented:
Ammended the runtime thanks to Mucker, made the lasers overlay over the machine, removed the (broken) plane handling, moved descriptions and added the boards for the checks to not freak out. Should be ready to re-review.

SierraKomodo commented on code/game/machinery/bluespace_drive.dm:
```diff
@@ -0,0 +1,145 @@
+	var/drive_integrity					// How much damage have we sustained?
+	var/drive_broken = FALSE			// Are we currently broken?
+	var/drive_repaired = FALSE			// Did someone repair us?
+	var/drive_core_unstable = FALSE		// Are we currently unstable?
```
Use dmdocs instead of line comments for these.

https://github.com/Baystation12/Baystation12/wiki/Guide-to-SpacemanDMM-Docblocks

SierraKomodo commented on code/game/machinery/bluespace_drive.dm:
```diff
@@ -0,0 +1,145 @@
+	update_icon()
+
+/obj/machinery/bluespacedrive/on_update_icon()
+	if(!drive_core_unstable)
+		overlays += "bsd_c_s"
+	else
+		overlays += "bsd_c_u"
```
This should run `overlays.Cut()` beforehand to clear out the existing overlays to prevent stacking and prevent a no longer relevant overlay from rendering with each update con call.

SierraKomodo commented on code/game/machinery/bluespace_drive.dm:
```diff
@@ -0,0 +1,145 @@
+				for(var/x in verbs)
+					verbs -= x
+				take_damage(drive_max_integrity)
+		if(3.0)
```
```suggestion
		if(1)
			qdel(src)
			return
		if(2)
			if (prob(25))
				qdel(src)
				return
			if (prob(50))
				for(var/x in verbs)
					verbs -= x
				take_damage(drive_max_integrity)
		if(3)
```

SierraKomodo commented on code/game/machinery/bluespace_drive.dm:
```diff
@@ -0,0 +1,145 @@
+			if (prob(50))
+				for(var/x in verbs)
+					verbs -= x
+				take_damage(drive_max_integrity)
```
You're removing verbs here but not in other cases of essentially killing the drive, i.e. in `emp_act()` - This feels inconsistent, is that intentional?

If not, it may be better to move the verb removal to the damage handling procs when any damage source 'kills' the drive instead.

SierraKomodo commented on code/game/machinery/bluespace_drive.dm:
```diff
@@ -0,0 +1,145 @@
+					verbs -= x
+				take_damage(drive_max_integrity)
+
+/obj/machinery/bluespacedrive/bullet_act(var/obj/item/projectile/Proj)
```
```suggestion
/obj/machinery/bluespacedrive/bullet_act(obj/item/projectile/Proj)
```

SierraKomodo commented on code/game/machinery/bluespace_drive.dm:
```diff
@@ -0,0 +1,145 @@
+	drive_integrity -= damage
+	if(drive_integrity <= 0)
+		playsound(loc, 'sound/machines/BSD_explosion.ogg', 100)
+		visible_message(SPAN_DANGER(FONT_LARGE("\The [src] beings emitting an ear-splitting, horrible shrill! get back!")))
```
```suggestion
		visible_message(SPAN_DANGER(FONT_LARGE("\The [src] begins emitting an ear-splitting, horrible shrill! Get back!")))
```

SierraKomodo commented on code/game/machinery/bluespace_drive.dm:
```diff
@@ -0,0 +1,145 @@
+	drive_core_unstable = TRUE
+	update_icon()
+
+/obj/machinery/bluespacedrive/attackby(var/obj/item/O, var/mob/user)
```
```suggestion
/obj/machinery/bluespacedrive/attackby(obj/item/O, mob/user)
```

SierraKomodo commented on code/game/machinery/bluespace_drive.dm:
```diff
@@ -0,0 +1,145 @@
+	. = ..()
+
+/obj/machinery/bluespacedrive/Process()
+	. = ..()
```
These are unnecessary if you're not actually adding anything to the overrides.

SierraKomodo commented on code/modules/research/designs/designs_circuits.dm:
```diff
@@ -826,6 +826,13 @@
+/datum/design/circuit/bluespacedrive
+	name = "Naophoros-pattern bluespace puncturer"
+	id = "bluespacedrive"
+	req_tech = list(TECH_BLUESPACE = 12, TECH_POWER = 12, TECH_ENGINEERING = 12) // You are not supposed to get this
```
If it's supposed to be unattainable why make a design datum for it?

MuckerMayhem replied:
There's a test that mandates machines have a circuit board for it so that they are "buildable", the circuit board here is purely to get that test to pass

SierraKomodo replied:
Silly arbitrary tests

SierraKomodo reviewed (changes requested):
Map stuff:

It feels a bit odd having the BSD sitting partially on top of the vent and scrubber. I suggest moving the vent/scrubber up/down a tile so they're not layered over each other.
![firefox_WRO97JVR1V](https://user-images.githubusercontent.com/11140088/168806989-230bcae7-33f7-4d6b-abaa-d3d51b765660.png)

SierraKomodo reviewed (dismissed):
And one other item regarding health handling.

If possible, please use the standardized health handlers as provided in `code\game\atoms_health.dm` instead of adding your own variant implementation.

SierraKomodo commented on code/game/machinery/bluespace_drive.dm:
```diff
@@ -0,0 +1,146 @@
+	var/drive_broken = FALSE		
+	var/drive_repaired = FALSE		
+	var/drive_core_unstable = FALSE		
+	var/drive_sound
```
This won't parse correctly for dmdocs - It'll consider the entire comment to be the documentation solely for `drive_max_integrity` and not apply it to any of the other vars.

```suggestion
	/// How much damage can it sustain
	var/drive_max_integrity = 1000		
	/// How much damage has it sustained
	var/drive_integrity			
	/// Is it currently broken?
	var/drive_broken = FALSE		
	/// Has it been repaired?
	var/drive_repaired = FALSE		
	/// Is the core currently unstable?
	var/drive_core_unstable = FALSE		
	var/drive_sound
```

SierraKomodo commented on code/game/machinery/bluespace_drive.dm:
```diff
@@ -0,0 +1,146 @@
+			if(501 to 750)
+				user.visible_message(SPAN_NOTICE("Its field is crackling gently, with the ocassional twitch."))
+			else
+				user.visible_message(SPAN_NOTICE("At a glance, its field is peacefully humming without any alterations."))
```
Also didn't catch this before, sorry - `user.visible_message(MESSAGE)` presents a message everyone near by can see instead of only posting a message to the user. For this you want to use `to_chat(user, MESSAGE)` instead.

SomeAngryMiner committed:
Tweaks to the BSD

CrimsonShrike committed:
TGmerged deck 2 remap (#4)

SomeAngryMiner commented:
Squashed, Mapmerged and all suggestions applied, with the exception of the atom health as discussed on the code channel.

Spookerton reviewed (approved):
As is, drive_repaired is unused and after the core explodes the bsd_c_u overlay remains and ambient sound continues to play. Nevertheless I'm happy to get this in and fix the little things afterward.
